[R-package][ci] added CI stage  for R package (fixes #2335, fixes #2569)

jameslamb opened the pull request:
I'm still committed to getting R CI running in GitHub Actions (#2353 ) but I think there is also an immediate benefit to having the R package built + unit tests run _at all_, and tacking that on to Travis is the easiest path.

In this PR, I propose adding an `r-pkg` stage to our Travis configuration to at least test the R package on Ubuntu and macOS. That will catch issues like #2524 , where a change on the C++ side breaks the R package. We can handle tests on Windows, tests with different versions of R, running on GitHub Actions, etc. in later PRs.

Opening as a draft PR since I may need to build a few times to get feedback from Travis. I'll change to regular PR and add reviewers when I think it's ready and worthy of your time, but comments are welcome any time!

jameslamb commented on .travis.yml:
```diff
@@ -6,24 +6,25 @@ git:
+    # - TASK=mpi METHOD=source
+    # - TASK=mpi METHOD=pip
+    # - TASK=gpu METHOD=source PYTHON_VERSION=3.5
+    # - TASK=gpu METHOD=pip PYTHON_VERSION=3.6
```
commenting out stuff to speed up the feedback cycle and minimize number of tasks running on Travis (since other PRs compete for those resources). Will revert once I feel this is in a working state

StrikerRUS commented:
I think that R environment from conda is not appropriate to run unit tests. Not many users use it, it has its own bugs, it has huge delay in updating package versions, [it doesn't work normally on Windows](https://github.com/microsoft/LightGBM/issues/1143#issuecomment-475920414), it means that compilation of LightGBM will be performed by [conda's specific compilers](https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html).

Can you please consider using vanilla R environment? https://github.com/microsoft/LightGBM/issues/1143#issuecomment-477563500

jameslamb commented:
> I think that R environment from conda is not appropriate to run unit tests. Not many users use it, it has its own bugs, it has huge delay in updating package versions, [it doesn't work normally on Windows](https://github.com/microsoft/LightGBM/issues/1143#issuecomment-475920414), it means that compilation of LightGBM will be performed by [conda's specific compilers](https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html).
> 
> Can you please consider using vanilla R environment? [#1143 (comment)](https://github.com/microsoft/LightGBM/issues/1143#issuecomment-477563500)

Sure I can look. To be clear, I'm only using conda as an installer for R itself (`r-base`) since we already had `conda` laying around. All the other packages I'm installing are source distributions directly from CRAN (those `install.packages()` calls).

I'll take a look at non-conda R environment.

jameslamb commented on R-package/R/lightgbm.R:
```diff
@@ -181,6 +181,7 @@ globalVariables(c(
     "."
     , ".N"
     , ".SD"
+    , "abs_contribution"
```
fixes `R CMD check` issue:

```
multiple.tree.interprete: no visible binding for global variable
  ‘abs_contribution’
Undefined global functions or variables:
  abs_contribution
```

jameslamb commented on R-package/R/lgb.cv.R:
```diff
@@ -359,7 +359,7 @@ lgb.cv <- function(params = list(),
       booster_old <- list(
         best_iter = fd$booster$best_iter
-        , best_score = fd$booster$best_score,
+        , best_score = fd$booster$best_score
```
bad comma, fixes `R CMD check` issue

```
lgb.cv : <anonymous>: missing arguments not allowed in calls to ‘list’
```

jameslamb commented on .ci/test.sh:
```diff
@@ -173,3 +173,73 @@ matplotlib.use\(\"Agg\"\)\
+
+if [[ $TASK == "r-pkg" ]]; then
+    #. set up R environment
+    R_LIB_PATH=~/Rlib
```
Borrowed your setup [from RGF](https://github.com/RGF-team/rgf/blob/master/R-package/.R.travis.sh) and adapted it slightly to (hopefully) avoid needing to build `devtools` and all its dependencies. Thanks @StrikerRUS !

jameslamb commented on .ci/test.sh:
```diff
@@ -173,3 +173,73 @@ matplotlib.use\(\"Agg\"\)\
+        sudo tlmgr update --self
+        # sudo tlmgr install \
+        #     inconsolata \
+        #     helvetic
```
will remove this commented code if it isn't necessary. Commented for now to test something

jameslamb commented:
I am getting this WARNING from `R CMD check`

![image](https://user-images.githubusercontent.com/7608904/67646200-dc1e5a80-f8fa-11e9-946e-6bcfc656e841.png)


Somehow additional files like a `.coveralls.yml` (which we don't even have in this project) and a `.git/` directory are being included in the package. I believe these are coming from the `compute` git submodule, so I need to figure out how to deal with that.

jameslamb commented:
ok next issue, I am seeing this [on Travis]() but not locally

```
* checking for GNU extensions in Makefiles ... WARNING
Found the following file(s) containing GNU extensions:
  src/build/Makefile
Portable Makefiles do not use GNU extensions such as +=, :=, $(shell),
$(wildcard), ifeq ... endif, .NOTPARALLEL See section ‘Writing portable
```

When I run locally, I do see that a file `lightgbm.Rcheck/00_pkg_src/build/Makefile` is created after running

```
export CXX=/usr/local/bin/g++-8
export CC=/usr/local/bin/gcc-8
Rscript build_r.R
R CMD check *.tar.gz --as-cran
```

However, that WARNING is not generated on my local machine (macOS High Sierra,R 3.5.0).

Of the extensions complained about in that warning, I only see one in that file:

```
# Allow only one "make -f Makefile" at a time, but pass parallelism.
.NOTPARALLEL:
```

Currently on Travis I'm building with R 3.6.1. I'm going to try with R3.5.0 just to see if the . R version difference is to blame.

Next I'll check `cmake` version differences, since this Makefile is something that is created by `cmake` (not a file we have checked into this package's source).

jameslamb commented on R-package/.Rbuildignore:
```diff
@@ -12,3 +12,14 @@
+^src/compute/.git$
+^src/compute/.gitignore$
+^src/compute/CONTRIBUTING.md$
+^src/compute/README.md$
```
`src/compute` is the boost stuff we pull in as a git submodule. It's important that the random other stuff from that repo isn't bundled with the R package, as `R CMD check` will complain about things like a `.git/` folder being put into a source distribution.

jameslamb commented on R-package/R/lgb.prepare_rules2.R:
```diff
@@ -46,7 +46,7 @@
 #'     "setosa" = 3L
 #'     , "versicolor" = 2L
-#'     , virginica" = 1L
+#'     , "virginica" = 1L
```
fixes

```
Error: LaTeX error. Mismatched quote or braces
```

When building the documentation.

jameslamb commented on .travis.yml:
```diff
@@ -14,16 +14,17 @@ env:
+    # - TASK=mpi METHOD=pip
+    # - TASK=gpu METHOD=source PYTHON_VERSION=3.5
+    # - TASK=gpu METHOD=pip PYTHON_VERSION=3.6
+    - TASK=r-pkg R_VERSION=3.6.1
```
right now we'll get this R version when building on Mac, but on Ubuntu we're at the whim of whatever is "latest" for the particular ubuntu version we are running. In this case the latest version for R available on bionic is 3.4.4 (see [here](https://packages.ubuntu.com/search?keywords=r-base)). You can see [here](https://cran.r-project.org/bin/linux/ubuntu/#installation) for more.

I think that ok is ok for now.

StrikerRUS replied:
Then this variable should be named something like `R_VERSION_MAC` to avoid any misleading interpretation.

BTW, why did you abandoned the idea to install the latest version from CRAN sources for Ubuntu?

Do we have any limitation in using chars here? 😄 `TASK=r-package` seems to be more pretty to see.

jameslamb replied:
ha nope no limitation, happy to use `r-package`. `r-pkg` is a bad habit of mine from other projects.

> why did you abandoned the idea to install the latest version from CRAN sources for Ubuntu

I followed the instructions from https://cran.r-project.org/bin/linux/ubuntu/README.html which SAY you should get 3.6.x, but still ended up with 3.4.x. I'm not sure why. We are using bionic (Ubuntu 18.04) on Travis, and it seems [3.4.x is the latest `r-base` release available](https://packages.ubuntu.com/search?keywords=r-base). I think I see the problem though...found [in this answer](https://stackoverflow.com/a/56378217/3986677) that there is one more step to see more versions from that server. I think in my latest commit, I have it successfully installing 3.6.1

jameslamb commented:
Alright, I think this PR is finally ready for review. I've left some comments above calling out items that I think might be concerning. I expect we'll have a conversation about two more things:

1. the `r-pkg` stage takes 10-11 minutes to run. I'm unsure how to get that down. I don't think it's TOO bad given what we get in return in terms of quality control
2. I had to put a gross hack into `install.libs.R` to deal with a new check in `R CMD check` for R3.6+. Basically `cmake` was generating a `Makefile` with `.NOTPARALLEL:` in it, which R complains about as not portable. My hack was to have `install.libs.R` rewrite that `Makefile` without `NOTPARALLEL` before executing it to build `lib_lightgbm`. I am open to other ideas!

@StrikerRUS I know you don't have an R environment to test some of the changes in, but I would definitely like your input on the changes I'm making to `.ci/test.sh` at a minimum (and of course anything else you notice 😀 )

StrikerRUS commented on .ci/test.sh:
```diff
@@ -75,6 +75,75 @@ if [[ $TASK == "if-else" ]]; then
     exit 0
 fi
 
+if [[ $TASK == "r-pkg" ]]; then
```
I feel like it can be moved into a separate `.ci/test_r_package.sh` file, since it's self-contained, rather big and includes a lot of installation stuff.

jameslamb replied:
ok sure, I'd be happy to do that. I wasn't sure where we draw the line between keeping `test.sh` readable and minimizing the number of files to manage in `.ci/`

StrikerRUS replied:
Yeah, it's not so clear... I proposed that mostly because there are a lot of preparation stuff in the script which we used to keep in `setup.sh` file.

jameslamb replied:
moved everything into `.ci/test_r_package.sh` in https://github.com/microsoft/LightGBM/pull/2530/commits/dc95c332739a709ff926e1589b0ecceaefc8ebe6

StrikerRUS commented on .ci/test.sh:
```diff
@@ -75,6 +75,11 @@ if [[ $TASK == "if-else" ]]; then
     exit 0
 fi
 
+if [[ $TASK == "r-package" ]]; then
```
Please try it to run on Azure Pipelines and with swapped compilers.

jameslamb replied:
Can you clarify this one @StrikerRUS ? I just added `r-package` task to `.vsts-ci.yml` in both Linux and macOS stages. Is that sufficient for this PR? Or would you like me to include setting up a windows environment in `test_windows.ps1`?

What do you mean by "with swapper compilers"? We use `gcc` for linux and `clang` for Mac already (https://github.com/microsoft/LightGBM/blob/master/.travis.yml#L44-L50).

StrikerRUS replied:
> Is that sufficient for this PR? Or would you like me to include setting up a windows environment in test_windows.ps1?

Sure! Windows part can be done in a separate PR, I guess.

> What do you mean by "with swapped compilers"?

Sorry for being not clear enough! I meant replace `gcc` with `clang` and vice versa everywhere and see what happens to make sure that R-package works with both compilers.

StrikerRUS replied:
I think we can pre-install R inside our docker for Azure Pipelines.
https://github.com/guolinke/lightgbm-ci-docker

jameslamb replied:
oh got it, cool!

I've created https://github.com/guolinke/lightgbm-ci-docker/pull/5 and then removed the "figure out what version of Ubuntu you are on" code in this PR as of https://github.com/microsoft/LightGBM/pull/2530/commits/1633830cb754527507dddc006a55715de391a3ab

jameslamb replied:
@StrikerRUS Right now, we're getting the following runs of `test_r_package.sh`:

* Travis - linux - `clang`
* Travis - macos - `gcc`
* Azure - linux - `gcc`
* Azure - macos - `clang`

That should be sufficient for 'test with swapped compilers', right?

StrikerRUS replied:
@jameslamb By `test with swapped compilers` I mean change all values to opposite ones.
If it was `clang`, it should be `gcc` and vice versa. Refer to https://github.com/guolinke/lightgbm-ci-docker/pull/3, https://github.com/microsoft/LightGBM/pull/2467/commits/2757b3f8b9dcd3d901485f2f14d99e7bfbbfdb7a as examples.

jameslamb replied:
ah you just mean "temporarily change them to get a build to be sure it works". Ok I can do that

StrikerRUS commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,70 @@
+# https://cran.r-project.org/bin/linux/ubuntu/#installation
+# adding steps from https://stackoverflow.com/a/56378217/3986677 to get latest version
+if [[ $OS_NAME == "linux" ]]; then
+    sudo apt-get update
```
I suspect `update` is not needed to be executed twice.

jameslamb replied:
hmm ok. I've tried removing it, let's see!

StrikerRUS commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,70 @@
+        --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9
+    sudo apt-get update
+    sudo apt-get install \
+        -y \
```
Add `--no-install-recommends`.

jameslamb replied:
added

jameslamb replied:
this is breaking things resulting in `Rscript: command not found`. I will try explicitly adding in `r-base-core` but if that doesn't work I think we should add it back in for now.

jameslamb replied:
and that is because I remove the `sudo installer` line accidentally! oy

jameslamb replied:
I see this error [on the linux side](https://travis-ci.org/microsoft/LightGBM/jobs/610688556?utm_medium=notification&utm_source=github_status):

```
W: GPG error: https://cloud.r-project.org/bin/linux/ubuntu bionic-cran35/ InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 51716619E084DAB9
E: The repository 'https://cloud.r-project.org/bin/linux/ubuntu bionic-cran35/ InRelease' is not signed.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
Executing: /tmp/apt-key-gpghome.Lt8WVNv2MN/gpg.1.sh --keyserver keyserver.ubuntu.com --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9
gpg: key 51716619E084DAB9: public key "Michael Rutter <marutter@gmail.com>" imported
gpg: Total number processed: 1
gpg:               imported: 1
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Version '3.6.1-3bionic' for 'r-base-dev' was not found
```

I think it is the result of removing the second `update`, so I believe we need to leave it.

StrikerRUS replied:
> I think it is the result of removing the second update, so I believe we need to leave it.

Yeah! It's quite expected. The second `update` should be left and the first one should be removed.

jameslamb replied:
hmmmm ok let me try

StrikerRUS commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,70 @@
+
+# Installing R precompiled for Mac OS 10.11 or higher
+if [[ $OS_NAME == "macos" ]]; then
+    wget https://cran.r-project.org/bin/macosx/R-${R_MAC_VERSION}.pkg -O R.pkg
```
Add `-q`.

jameslamb replied:
added

StrikerRUS commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,70 @@
+fi
+
+conda install \
+    -y \
```
Add `-q`.

jameslamb replied:
added

StrikerRUS commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,70 @@
+if [[ $OS_NAME == "macos" ]]; then
+    wget https://cran.r-project.org/bin/macosx/R-${R_MAC_VERSION}.pkg -O R.pkg
+    sudo installer \
+        -verbose \
```
Is it really needed?

jameslamb replied:
I always like verbose logs for the purpose of CI, but I've turned this off to be consistent with how the rest of `test.sh` is structured

StrikerRUS replied:
Ah, I see! But at some point, with high logging level logs become so huge and in consequence poorly readable that cannot be rendered at CI web interface and you cannot find needed info quickly 😢 .

jameslamb commented:
@StrikerRUS please see https://github.com/microsoft/LightGBM/pull/2530/commits/64a46b5abcb37d21227cb45983cca4ed0271c0d1 and let me know what you think. Azure pipelines was failing because we are currently getting Ubuntu 14.04 (`trusty`)  and I had hard-coded `test_r_package.sh` for Ubuntu 18.04 (`bionic`) since that's what we get on Travis.

I think the solution in that commit should work.

StrikerRUS commented:
@jameslamb 
> please see 64a46b5 and let me know what you think. 

I think that taking into account this my https://github.com/microsoft/LightGBM/pull/2530#discussion_r345181493, after pre-installing R into docker there will be no need to install it inside the `test_r_package.sh` script.

jameslamb commented on .vsts-ci.yml:
```diff
@@ -24,24 +24,27 @@ jobs:
+      # gpu_source:
+      #   TASK: gpu
+      #   METHOD: source
+      #   PYTHON_VERSION: 3.6
```
I am testing an Azure-specific thing, trying to shorten the feedback cycle. Will be sure to uncomment these once I have it working.

jameslamb commented:
woo Linux build on Azure is now working! https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=3552

thanks to https://github.com/guolinke/lightgbm-ci-docker/pull/5

now to fix the libomp issues on mac

getting closer!

jameslamb commented:
I'm pretty confused why the Mac builds on Azure are failing. I'm getting this error (an old friend):

```
OMP: Error #15: Initializing libomp.dylib, but found libomp.dylib already initialized.
```

But this really should be fixed already by this

https://github.com/microsoft/LightGBM/blob/master/.ci/test.sh#L83-L86

jameslamb commented:
> I'm pretty confused why the Mac builds on Azure are failing. I'm getting this error (an old friend):
> 
> ```
> OMP: Error #15: Initializing libomp.dylib, but found libomp.dylib already initialized.
> ```
> 
> But this really should be fixed already by this
> 
> https://github.com/microsoft/LightGBM/blob/master/.ci/test.sh#L83-L86

found it!! We're getting another `libomp.dylib` from R!

```
/Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libomp.dylib
```

jameslamb commented:
ok @Laurae2 @guolinke @StrikerRUS I think I have this working and have addressed all comments! I've put a few of the smaller, noncontroversial changes in #2581 so if that gets merged I can rebase to make the diff here a bit more focused.

Looking forward to your thoughts. Hopefully we are close!

StrikerRUS commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,82 @@
+        sudo ln -sf \
+            "$(brew --cellar libomp)"/*/lib/libomp.dylib \
+            /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libomp.dylib
+    fi
```
I think it's better to add `... || exit -1` than `echo`. Also, my assumption is that this workaround is needed only for `$COMPILER == "clang"`.

jameslamb replied:
ah I actually only had this `echo` in there for debugging! I actually meant to remove it.

I think you are right that it's only needing for `clang`, I can add that guard.

StrikerRUS commented:
One more thing that we need to do is to think about CI **load balancing**. At present, we have roughly all CI jobs (Appveyor, Travis, Azure) finish in 15-17 minutes. Travis seems to have no limits at free plan for concurrent jobs exclusively for Microsoft repos, while Azure have the limit 10 jobs. Appveyor have no concurrent jobs at all. Perfectly, we should preserve the situation in which all CIs finish in the same time. 
I think that if we want to test at Azure **and** Travis, we should bring some **diversity** in testing environment, like we are doing now. Maybe different R versions make sense? Or different compilers is enough?

cc @guolinke @chivee

jameslamb commented:
> One more thing that we need to do is to think about CI **load balancing**. At present, we have roughly all CI jobs (Appveyor, Travis, Azure) finish in 15-17 minutes. Travis seems to have no limits at free plan for concurrent jobs exclusively for Microsoft repos, while Azure have the limit 10 jobs. Appveyor have no concurrent jobs at all. Perfectly, we should preserve the situation in which all CIs finish in the same time.
> I think that if we want to test at Azure **and** Travis, we should bring some **diversity** in testing environment, like we are doing now. Maybe different R versions make sense? Or different compilers is enough?
> 
> **UPD:** Do we need to provide any artifacts for R-package?
> 
> cc @guolinke @chivee

I am open to adding some diversity but I also would like to kindly request that that decision be moved to a discussion on an issue and outside of this PR. Right now someone can make a PR that breaks the R package in a serious way but does not fail CI. I would like to prevent that as soon as possible, before we talk about optimizations on the CI side of expansion of what is tested.

jameslamb commented:
I just rebased this to most recent `master` and added https://github.com/microsoft/LightGBM/pull/2530/commits/6ca9091093af3cd92890979093e4847008f0ae38 which tries swapping `gcc` and `clang` everywhere in Azure CI and Travis. Let's see if it builds as expected.

@StrikerRUS could you let me know what you think about https://github.com/microsoft/LightGBM/pull/2530#issuecomment-556927013?

re: the question in https://github.com/microsoft/LightGBM/pull/2530#issuecomment-556004406 ... I don't think we need to produce artifacts for the R package as part of this PR. That could be something done in a later PR I suppose. I really would like to have a working CI environment set up for R soon, as I think it's important to be sure changes like #2594 and #2567 don't have any unintended side effects on the R package.

StrikerRUS commented:
@jameslamb 
> could you let me know what you think about #2530 (comment)?

I'm OK with your position to merge this PR ASAP, because it's a one more step towards so long-awaited CRAN release. Let's spectate for CI timings with R-tasks and think about re-balancing jobs later (if needed).

> I just rebased this to most recent master and added 6ca9091 which tries swapping gcc and clang everywhere in Azure CI and Travis. Let's see if it builds as expected.

```
  OMP: Error #15: Initializing libomp.dylib, but found libomp.dylib already initialized.
  OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program. That is dangerous, since it can degrade performance or cause incorrect results. The best thing to do is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding static linking of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented workaround you can set the environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, but that may cause crashes or silently produce incorrect results. For more information, please see http://openmp.llvm.org/
```

Seems there is one more libomp instance somewhere!..

jameslamb commented:
> UPD: I guess removing $AZURE == "true" condition will help.

Thanks @StrikerRUS you were right. Seems like that makes things work again. Posting links below for if / when we reference this comment in the future.

With the compilers flipped (https://github.com/microsoft/LightGBM/pull/2530/commits/eb927c9cbf4866afee84fd6b8fee8f109d4ad58f)

![image](https://user-images.githubusercontent.com/7608904/70022234-22e61c80-1559-11ea-97d5-fc0d0caec961.png)

Seems everything is working!

* [Azure devops](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=3683)
* [AppVeyor](https://ci.appveyor.com/project/guolinke/lightgbm/builds/29268986)
* [Travis](https://travis-ci.org/microsoft/LightGBM/builds/619951053?utm_source=github_status&utm_medium=notification)

Want me to switch the compilers back before we merge?

StrikerRUS commented on R-package/src/install.libs.R:
```diff
@@ -99,6 +99,32 @@ if (!use_precompile) {
+
+  # R CMD check complains about the .NOTPARALLEL directive created in the cmake
+  # Makefile. We don't need it here anyway since targets are built serially, so trying
+  # to remove it with this hack
```
Is it safe? I remember setting `-j4` allowed to speed up building significantly. Forgive me if this hack is not about that.
https://github.com/microsoft/LightGBM/blob/6129208006c91f9ac94891690aca6f0c49d9b08e/.ci/test.sh#L151

jameslamb replied:
I don't know why the `Makefile` being generated sometimes has the `.NOTPARALLEL` directive in it, but I know that that is non-portable which causes `R CMD CHECK` to complain. Removing this might make using parallelization with `-j4` unsafe, but I think it is ok. Right now with the `Rscript build_r.R` process, we are not giving users any ability to customize the command e.g. to choose `make _lightgbm` vs. `make _lightgbm -j4`. I think it's ok to hack around this to get us closer to CRAN, and come back to it if / when we introduce a build process for R that uses parallelization when creating `liblightgbm`.

StrikerRUS replied:
Agree!

jameslamb commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,81 @@
+Rscript -e "install.packages(c('data.table', 'jsonlite', 'Matrix', 'R6', 'testthat'))" || exit -1
+
+cd ${BUILD_DIRECTORY}
+Rscript build_r.R
```
makes sense, I can do that

jameslamb commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,81 @@
+# R CMD CHECK
+R CMD check ${PKG_TARBALL} \
+    --as-cran \
+    --no-manual \
```
you are right! I must have added this at some point during debugging and forgot to remove it. Let me try right now 😬

jameslamb commented on .vsts-ci.yml:
```diff
@@ -84,6 +86,9 @@ jobs:
         TASK: bdist
+      r_package:
+        TASK: r-package
+        R_MAC_VERSION: 3.6.1
```
makes sense! I can do that

jameslamb commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,81 @@
+    if [[ $COMPILER == "clang" ]]; then
+        sudo ln -sf \
+            "$(brew --cellar libomp)"/*/lib/libomp.dylib \
+            /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libomp.dylib
```
yes you are right!  It's very annoying that they use `/major.minor` but all the built-ins for R return `major` and `minor.patch`, so you have to parse it apart.

I had to do this :/

```
R_MAJOR_MINOR=$(
    Rscript -e 'cat(paste0(R.version[["major"]], ".", strsplit(R.version[["minor"]], "\\\.")[[1]][1]))'
)
```

jameslamb replied:
To be clear, I didn't not do this because it's hard...you're right that I overlooked this, and THEN trying to fix it I realized that it's hard 😆

StrikerRUS replied:
Sorry, didn't get it. Why did you use R? I think we are in a shell script and simple version parsing is enough:
```
ver_arr=( ${R_MAC_VERSION//./ } )
```

![image](https://user-images.githubusercontent.com/25141164/70104639-b69a1480-164f-11ea-95f7-e4d2e22fee9a.png)

jameslamb replied:
oh wow yes I like this better. I actually did not know the `//./ `  syntax

jameslamb commented:
Now that I took out `--no-manual`, I'm seeing issues.

On [Linux](https://travis-ci.org/microsoft/LightGBM/jobs/620315392?utm_medium=notification&utm_source=github_status):

```
* checking PDF version of manual ... WARNING
LaTeX errors when creating PDF version.
This typically indicates Rd problems.
```

On [Mac](https://travis-ci.org/microsoft/LightGBM/jobs/620315399?utm_medium=notification&utm_source=github_status)

```
* checking PDF version of manual ... WARNING
LaTeX errors when creating PDF version.
This typically indicates Rd problems.
* checking PDF version of manual without hyperrefs or index ... ERROR
Re-running with no redirection of stdout/stderr.
Hmm ... looks like a package
Error in texi2dvi(file = file, pdf = TRUE, clean = clean, quiet = quiet,  : 
  pdflatex is not available
Error in texi2dvi(file = file, pdf = TRUE, clean = clean, quiet = quiet,  : 
  pdflatex is not available
Error in running tools::texi2pdf()
You may want to clean up by 'rm -Rf /var/folders/17/5mc7816d3mndxjqgplq6057w0000gn/T//RtmpeI5Jhs/Rd2pdf621545fd1421'
* DONE
```

Will see if I can reproduce locally.

jameslamb commented:
Still struggling with LaTeX stuff now that I removed `--no-manual`. Will push a few more commits (I'm not able to replicate locally :/) and comment when I think it's working again.

StrikerRUS commented:
@jameslamb For macOS,
```
* checking PDF version of manual ... WARNING
LaTeX errors when creating PDF version.
This typically indicates Rd problems.
* checking PDF version of manual without hyperrefs or index ... ERROR
Re-running with no redirection of stdout/stderr.
Hmm ... looks like a package
Error in texi2dvi(file = file, pdf = TRUE, clean = clean, quiet = quiet,  : 
  pdflatex is not available
Error in texi2dvi(file = file, pdf = TRUE, clean = clean, quiet = quiet,  : 
  pdflatex is not available
Error in running tools::texi2pdf()
You may want to clean up by 'rm -Rf /var/folders/17/5mc7816d3mndxjqgplq6057w0000gn/T//RtmpeI5Jhs/Rd2pdf621545fd1421'
* DONE
```

I think you should install `qpdf`. Refer to 
https://github.com/RGF-team/rgf/blob/44b3654b5801273bc1a2e0d291a734f43d6b0b48/R-package/.R.travis.sh#L25

And probably some other tweaks, that you can find on next lines.

StrikerRUS commented:
@jameslamb 
> Will push a few more commits (I'm not able to replicate locally :/) and comment when I think it's working again.

You can create a new branch and test your assumptions there. By doing so, PR will have only clean and verified commits without long history. The only (dis)advantage is that nobody will see your your sufferings. 😄

jameslamb commented:
> @jameslamb
> 
> > Will push a few more commits (I'm not able to replicate locally :/) and comment when I think it's working again.
> 
> You can create a new branch and test your assumptions there. By doing so, PR will have only clean and verified commits without long history. The only (dis)advantage is that nobody will see your your sufferings. 😄

sure no problem. I figured that wasn't too important since we squash on merge anyway. But yeah I'll go test on a separate branch.

jameslamb commented:
Ok we're getting really close! I got the LaTeX stuff fixed on everything except linux builds for Azure DevOps. I needed to add a `yum install texinfo`. @StrikerRUS could you take a look at https://github.com/guolinke/lightgbm-ci-docker/pull/6

StrikerRUS commented on .travis.yml:
```diff
@@ -24,6 +24,7 @@ env:
     - TASK=mpi METHOD=pip
     - TASK=gpu METHOD=source PYTHON_VERSION=3.5
     - TASK=gpu METHOD=pip PYTHON_VERSION=3.6
+    - TASK=r-package R_MAC_VERSION=3.6.1 R_TRAVIS_LINUX_VERSION=3.6.1-3bionic
```
The same about `R_MAC_VERSION` and `R_TRAVIS_LINUX_VERSION` as in https://github.com/microsoft/LightGBM/pull/2530#discussion_r353193184.

Place to define them for Travis:
https://github.com/microsoft/LightGBM/blob/69c1c33093cb9fa1afd896b8a458e47aa2930c92/.travis.yml#L39-L58

jameslamb replied:
ah I see, sure no problem I can do that

StrikerRUS commented on docs/conf.py:
```diff
@@ -234,7 +234,7 @@ def generate_r_docs(app):
     /home/docs/.conda/bin/conda install -q -y -n r_env -c conda-forge \
         r-pkgdown=1.3.0=r351h6115d3f_1000 \
-        r-roxygen2=6.1.1=r35h0357c0b_1001
+        r-roxygen2>=7.0.2
```
Is it supposed to be fixed now? Then I should test in at RTD.
BTW, why `>=`? Are LightGBM's docs now incompatible with <7.0.2?

jameslamb replied:
yes, it's up to version 7.0.2 on `conda-forge` now. I _believe_ that as of the changes in this PR LightGBM's docs will be compatible with both 6.x and 7.x, but 7.0.2 is the newest release so I thought `<=` made sense just to be sure we don't get older versions.

If you think it's not necessary I'd be happy to revert this change.

StrikerRUS replied:
I just wondered about the current state of compatibility 🙂  I personally interpret `>=` in requirements as incompatibility for older versions. However, I'm OK with either `>=` or complete removal of version specification (as it should be compatible with all versions).

Setting docs test right now to check.

StrikerRUS replied:
Unfortunately it fails. So let's adjust versions for RTD in a separate PR, if you don't mind.

```
[autosummary] generating autosummary for: /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.Booster.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.Dataset.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.LGBMClassifier.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.LGBMModel.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.LGBMRanker.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.LGBMRegressor.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.create_tree_digraph.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.cv.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.early_stopping.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.plot_importance.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.plot_metric.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.plot_split_value_histogram.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.plot_tree.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.print_evaluation.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.record_evaluation.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.reset_parameter.rst, /home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/pythonapi/lightgbm.train.rst

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/conf.py", line 256, in generate_r_docs
    raise RuntimeError(output)
RuntimeError: Collecting package metadata: ...working... done
Solving environment: ...working... done

## Package Plan ##

  environment location: /home/docs/.conda/envs/r_env

  added / updated specs:
    - cmake==3.14.0=h52cb24c_0
    - r-base==3.5.1=h1e0a451_2
    - r-data.table==1.11.4=r351h96ca727_0
    - r-devtools==1.13.6=r351h6115d3f_0
    - r-jsonlite==1.5=r351h96ca727_0
    - r-matrix==1.2_14=r351h96ca727_0
    - r-testthat==2.0.0=r351h29659fb_0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    _libgcc_mutex-0.1          |             main           3 KB
    _r-mutex-1.0.0             |      anacondar_1           2 KB
    binutils_impl_linux-64-2.28.1|       had2808c_3        16.1 MB
    binutils_linux-64-7.2.0    |      had2808c_27           8 KB
    bwidget-1.9.11             |                1         113 KB
    bzip2-1.0.8                |       h7b6447c_0         105 KB
    ca-certificates-2019.11.27 |                0         132 KB
    cairo-1.14.12              |       h8948797_3         1.3 MB
    cmake-3.14.0               |       h52cb24c_0         9.5 MB
    curl-7.67.0                |       hbc83047_0         145 KB
    expat-2.2.6                |       he6710b0_0         187 KB
    fontconfig-2.13.0          |       h9420a91_0         291 KB
    freetype-2.9.1             |       h8a8886c_1         822 KB
    fribidi-1.0.5              |       h7b6447c_0         112 KB
    gcc_impl_linux-64-7.2.0    |       habb00fd_3        72.4 MB
    gcc_linux-64-7.2.0         |      h550dcbe_27           9 KB
    gfortran_impl_linux-64-7.2.0|       hdf63c60_3         8.9 MB
    gfortran_linux-64-7.2.0    |      h550dcbe_27           9 KB
    glib-2.56.2                |       hd408876_0         5.0 MB
    graphite2-1.3.13           |       h23475e2_0         101 KB
    gsl-2.4                    |       h14c3975_4         2.6 MB
    gxx_impl_linux-64-7.2.0    |       hdf63c60_3        18.6 MB
    gxx_linux-64-7.2.0         |      h550dcbe_27           8 KB
    harfbuzz-1.8.8             |       hffaf4a1_0         863 KB
    icu-58.2                   |       h9c2bf20_1        22.5 MB
    jpeg-9b                    |       h024ee3a_2         248 KB
    krb5-1.16.1                |       h173b8e3_7         1.4 MB
    libcurl-7.67.0             |       h20c2e04_0         576 KB
    libgcc-ng-9.1.0            |       hdf63c60_0         8.1 MB
    libgfortran-ng-7.3.0       |       hdf63c60_0         1.3 MB
    libpng-1.6.37              |       hbc83047_0         364 KB
    libssh2-1.8.2              |       h1ba5d50_0         250 KB
    libstdcxx-ng-9.1.0         |       hdf63c60_0         4.0 MB
    libtiff-4.1.0              |       h2733197_0         589 KB
    libuuid-1.0.3              |       h1bed415_2          16 KB
    libxcb-1.13                |       h1bed415_1         502 KB
    libxml2-2.9.9              |       hea5a465_1         2.0 MB
    make-4.2.1                 |       h1bed415_1         429 KB
    openssl-1.1.1d             |       h7b6447c_3         3.7 MB
    pango-1.42.4               |       h049681c_0         528 KB
    pcre-8.43                  |       he6710b0_0         260 KB
    pixman-0.38.0              |       h7b6447c_0         618 KB
    r-assertthat-0.2.0         |   r351h6115d3f_0          57 KB
    r-base-3.5.1               |       h1e0a451_2        39.8 MB
    r-cli-1.0.0                |   r351h6115d3f_1         323 KB
    r-crayon-1.3.4             |   r351h6115d3f_0         745 KB
    r-curl-3.2                 |   r351hadc6856_1         406 KB
    r-data.table-1.11.4        |   r351h96ca727_0         1.4 MB
    r-devtools-1.13.6          |   r351h6115d3f_0         684 KB
    r-digest-0.6.15            |   r351h96ca727_0         150 KB
    r-git2r-0.23.0             |   r351h96ca727_1         863 KB
    r-httr-1.3.1               |   r351h6115d3f_1         437 KB
    r-jsonlite-1.5             |   r351h96ca727_0         1.1 MB
    r-lattice-0.20_35          |   r351h96ca727_0         1.1 MB
    r-magrittr-1.5             |   r351h6115d3f_4         165 KB
    r-matrix-1.2_14            |   r351h96ca727_0         3.7 MB
    r-memoise-1.1.0            |   r351h6115d3f_0          40 KB
    r-mime-0.5                 |   r351h96ca727_0          38 KB
    r-openssl-1.0.2            |   r351h96ca727_1         1.1 MB
    r-praise-1.0.0             |   r351h6115d3f_4          21 KB
    r-r6-2.2.2                 |   r351h6115d3f_0         275 KB
    r-rlang-0.2.1              |   r351h96ca727_0         776 KB
    r-rstudioapi-0.7           |   r351h6115d3f_0         132 KB
    r-testthat-2.0.0           |   r351h29659fb_0         960 KB
    r-whisker-0.3_2            |   r351hf348343_4          66 KB
    r-withr-2.1.2              |   r351h6115d3f_0         161 KB
    rhash-1.3.8                |       h1ba5d50_0         235 KB
    tktable-2.10               |       h14c3975_0          88 KB
    xz-5.2.4                   |       h14c3975_4         366 KB
    zstd-1.3.7                 |       h0b5b093_0         887 KB
    ------------------------------------------------------------
                                           Total:       240.4 MB

The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main
  _r-mutex           pkgs/r/linux-64::_r-mutex-1.0.0-anacondar_1
  binutils_impl_lin~ pkgs/main/linux-64::binutils_impl_linux-64-2.28.1-had2808c_3
  binutils_linux-64  pkgs/main/linux-64::binutils_linux-64-7.2.0-had2808c_27
  bwidget            pkgs/main/linux-64::bwidget-1.9.11-1
  bzip2              pkgs/main/linux-64::bzip2-1.0.8-h7b6447c_0
  ca-certificates    pkgs/main/linux-64::ca-certificates-2019.11.27-0
  cairo              pkgs/main/linux-64::cairo-1.14.12-h8948797_3
  cmake              pkgs/main/linux-64::cmake-3.14.0-h52cb24c_0
  curl               pkgs/main/linux-64::curl-7.67.0-hbc83047_0
  expat              pkgs/main/linux-64::expat-2.2.6-he6710b0_0
  fontconfig         pkgs/main/linux-64::fontconfig-2.13.0-h9420a91_0
  freetype           pkgs/main/linux-64::freetype-2.9.1-h8a8886c_1
  fribidi            pkgs/main/linux-64::fribidi-1.0.5-h7b6447c_0
  gcc_impl_linux-64  pkgs/main/linux-64::gcc_impl_linux-64-7.2.0-habb00fd_3
  gcc_linux-64       pkgs/main/linux-64::gcc_linux-64-7.2.0-h550dcbe_27
  gfortran_impl_lin~ pkgs/main/linux-64::gfortran_impl_linux-64-7.2.0-hdf63c60_3
  gfortran_linux-64  pkgs/main/linux-64::gfortran_linux-64-7.2.0-h550dcbe_27
  glib               pkgs/main/linux-64::glib-2.56.2-hd408876_0
  graphite2          pkgs/main/linux-64::graphite2-1.3.13-h23475e2_0
  gsl                pkgs/main/linux-64::gsl-2.4-h14c3975_4
  gxx_impl_linux-64  pkgs/main/linux-64::gxx_impl_linux-64-7.2.0-hdf63c60_3
  gxx_linux-64       pkgs/main/linux-64::gxx_linux-64-7.2.0-h550dcbe_27
  harfbuzz           pkgs/main/linux-64::harfbuzz-1.8.8-hffaf4a1_0
  icu                pkgs/main/linux-64::icu-58.2-h9c2bf20_1
  jpeg               pkgs/main/linux-64::jpeg-9b-h024ee3a_2
  krb5               pkgs/main/linux-64::krb5-1.16.1-h173b8e3_7
  libcurl            pkgs/main/linux-64::libcurl-7.67.0-h20c2e04_0
  libedit            pkgs/main/linux-64::libedit-3.1.20181209-hc058e9b_0
  libffi             pkgs/main/linux-64::libffi-3.2.1-hd88cf55_4
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0
  libgfortran-ng     pkgs/main/linux-64::libgfortran-ng-7.3.0-hdf63c60_0
  libpng             pkgs/main/linux-64::libpng-1.6.37-hbc83047_0
  libssh2            pkgs/main/linux-64::libssh2-1.8.2-h1ba5d50_0
  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0
  libtiff            pkgs/main/linux-64::libtiff-4.1.0-h2733197_0
  libuuid            pkgs/main/linux-64::libuuid-1.0.3-h1bed415_2
  libxcb             pkgs/main/linux-64::libxcb-1.13-h1bed415_1
  libxml2            pkgs/main/linux-64::libxml2-2.9.9-hea5a465_1
  make               pkgs/main/linux-64::make-4.2.1-h1bed415_1
  ncurses            pkgs/main/linux-64::ncurses-6.1-he6710b0_1
  openssl            pkgs/main/linux-64::openssl-1.1.1d-h7b6447c_3
  pango              pkgs/main/linux-64::pango-1.42.4-h049681c_0
  pcre               pkgs/main/linux-64::pcre-8.43-he6710b0_0
  pixman             pkgs/main/linux-64::pixman-0.38.0-h7b6447c_0
  r-assertthat       pkgs/r/linux-64::r-assertthat-0.2.0-r351h6115d3f_0
  r-base             pkgs/r/linux-64::r-base-3.5.1-h1e0a451_2
  r-cli              pkgs/r/linux-64::r-cli-1.0.0-r351h6115d3f_1
  r-crayon           pkgs/r/linux-64::r-crayon-1.3.4-r351h6115d3f_0
  r-curl             pkgs/r/linux-64::r-curl-3.2-r351hadc6856_1
  r-data.table       pkgs/r/linux-64::r-data.table-1.11.4-r351h96ca727_0
  r-devtools         pkgs/r/linux-64::r-devtools-1.13.6-r351h6115d3f_0
  r-digest           pkgs/r/linux-64::r-digest-0.6.15-r351h96ca727_0
  r-git2r            pkgs/r/linux-64::r-git2r-0.23.0-r351h96ca727_1
  r-httr             pkgs/r/linux-64::r-httr-1.3.1-r351h6115d3f_1
  r-jsonlite         pkgs/r/linux-64::r-jsonlite-1.5-r351h96ca727_0
  r-lattice          pkgs/r/linux-64::r-lattice-0.20_35-r351h96ca727_0
  r-magrittr         pkgs/r/linux-64::r-magrittr-1.5-r351h6115d3f_4
  r-matrix           pkgs/r/linux-64::r-matrix-1.2_14-r351h96ca727_0
  r-memoise          pkgs/r/linux-64::r-memoise-1.1.0-r351h6115d3f_0
  r-mime             pkgs/r/linux-64::r-mime-0.5-r351h96ca727_0
  r-openssl          pkgs/r/linux-64::r-openssl-1.0.2-r351h96ca727_1
  r-praise           pkgs/r/linux-64::r-praise-1.0.0-r351h6115d3f_4
  r-r6               pkgs/r/linux-64::r-r6-2.2.2-r351h6115d3f_0
  r-rlang            pkgs/r/linux-64::r-rlang-0.2.1-r351h96ca727_0
  r-rstudioapi       pkgs/r/linux-64::r-rstudioapi-0.7-r351h6115d3f_0
  r-testthat         pkgs/r/linux-64::r-testthat-2.0.0-r351h29659fb_0
  r-whisker          pkgs/r/linux-64::r-whisker-0.3_2-r351hf348343_4
  r-withr            pkgs/r/linux-64::r-withr-2.1.2-r351h6115d3f_0
  readline           pkgs/main/linux-64::readline-7.0-h7b6447c_5
  rhash              pkgs/main/linux-64::rhash-1.3.8-h1ba5d50_0
  tk                 pkgs/main/linux-64::tk-8.6.8-hbc83047_0
  tktable            pkgs/main/linux-64::tktable-2.10-h14c3975_0
  xz                 pkgs/main/linux-64::xz-5.2.4-h14c3975_4
  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3
  zstd               pkgs/main/linux-64::zstd-1.3.7-h0b5b093_0


Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done

Error: package or namespace load failed for ‘pkgdown’ in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]):
 namespace ‘purrr’ 0.2.5 is being loaded, but >= 0.3.3 is required
Execution halted


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/docs/checkouts/readthedocs.org/user_builds/lightgbm/envs/docs/lib/python3.7/site-packages/sphinx/cmd/build.py", line 303, in build_main
    args.tags, args.verbosity, args.jobs, args.keep_going)
  File "/home/docs/checkouts/readthedocs.org/user_builds/lightgbm/envs/docs/lib/python3.7/site-packages/sphinx/application.py", line 263, in __init__
    self._init_builder()
  File "/home/docs/checkouts/readthedocs.org/user_builds/lightgbm/envs/docs/lib/python3.7/site-packages/sphinx/application.py", line 325, in _init_builder
    self.emit('builder-inited')
  File "/home/docs/checkouts/readthedocs.org/user_builds/lightgbm/envs/docs/lib/python3.7/site-packages/sphinx/application.py", line 510, in emit
    return self.events.emit(event, self, *args)
  File "/home/docs/checkouts/readthedocs.org/user_builds/lightgbm/envs/docs/lib/python3.7/site-packages/sphinx/events.py", line 80, in emit
    results.append(callback(*args))
  File "/home/docs/checkouts/readthedocs.org/user_builds/lightgbm/checkouts/docs/docs/conf.py", line 260, in generate_r_docs
    raise Exception("An error has occurred while generating documentation for R-package\n" + str(e))
Exception: An error has occurred while generating documentation for R-package
Collecting package metadata: ...working... done
Solving environment: ...working... done
```

jameslamb replied:
ah ok, I'll take it out of this PR

jameslamb replied:
reverted in https://github.com/microsoft/LightGBM/pull/2530/commits/d98211ff6475b97e77b95e03743c850b00dde055

StrikerRUS commented on .ci/test_r_package.sh:
```diff
@@ -0,0 +1,90 @@
+cd ${BUILD_DIRECTORY}
+Rscript build_r.R || exit -1
+
+PKG_TARBALL=$(ls | grep '^lightgbm_.*\.tar\.gz$')
```
You can use `LGB_VER` environment variable for the specifics.

jameslamb replied:
ah I did not know about that, ok great

jameslamb commented:
Think I've addressed the most recent batch of comments in https://github.com/microsoft/LightGBM/pull/2530/commits/1f5f339c489ae8478e5a216fb0612b236a9f1b1e

That is also the first build using the new container on `dev` of https://github.com/guolinke/lightgbm-ci-docker and [looks like it's working](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=3746)!

StrikerRUS commented:
@jameslamb 
> That is also the first build using the new container on dev of https://github.com/guolinke/lightgbm-ci-docker and looks like it's working!

Very awesome! Please switch compilers back to their original values.

jameslamb commented:
> @jameslamb
> 
> > That is also the first build using the new container on dev of https://github.com/guolinke/lightgbm-ci-docker and looks like it's working!
> 
> Very awesome! Please switch compilers back to their original values.

thanks! just switched them back in https://github.com/microsoft/LightGBM/pull/2530/commits/d98211ff6475b97e77b95e03743c850b00dde055

jameslamb commented:
All working!  @StrikerRUS could you take one more look?

StrikerRUS commented on .vsts-ci.yml:
```diff
@@ -11,7 +11,7 @@ variables:
   containers:
   - container: ubuntu1404
-    image: lightgbm/vsts-agent:ubuntu-14.04
+    image: lightgbm/vsts-agent:ubuntu-14.04-dev
```
@jameslamb 
> All working! @StrikerRUS could you take one more look?

👍  

https://github.com/guolinke/lightgbm-ci-docker/pull/7 has been merged, so you can switch container back to the production version.

jameslamb replied:
switched back in https://github.com/microsoft/LightGBM/pull/2530/commits/9ab619c955e96321c474cc5f37ffce6687d6c368

jameslamb commented:
> LGTM! Thanks!

Thanks so much! I think it's ok to merge with your approval + @Laurae2 's. Do you agree? This one touches a lot of stuff so want to be sure it's ok.

StrikerRUS commented:
@jameslamb Sure! @Laurae2's review was outdated, so I re-requested a fresh one from him.

jameslamb commented:
> @jameslamb Sure! @Laurae2's review was outdated, so I re-requested a fresh one from him.

@Laurae2 could you take one more look?

jameslamb commented:
Thanks for the reviews and help along the way! I'm excited to merge this one 😀
